src/navigation_table.py

```diff
@@ -0,0 +1,129 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Class for parsing and rendering a navigation table."""
```

Comment:
I think https://github.com/Python-Markdown/markdown is a good choice.

Reply:
Yeh I looked at those, usually they are meant for converting to another format like HTML rather than giving programmatic access to the values within a table, for example. For example for the markdown library:

```Python
import markdown
>>> markdown.markdown("# title 1\n\nbody")
'<h1>title 1</h1>\n<p>body</p>'
```

Which is not really what is needed here, the code just needs to extract key information from the table and re-generate an updated table at the end. Another option would be to convert to HTML and then parse it, that just seems to get pretty convoluted for something that is reasonably straight forward to do with a few regular expressions. The table will usually be machine generated and interpreted, so the code controls both the writing and reading of the table except on initial use where a table could have been hand written like in the case of indico. The cost of a mistake in reading that table is minimal since the table just gets re-generated based on the file system anyway.

Reply:
The library also doesn't seem to do cleanup (see the `href` being generated here):

```Python
import markdown
>>> markdown.markdown("|1|a|[b ]()|")
'<p>|1|a|<a href="">b </a>|</p>'
```

Reply:
I think this lib is better suited for you needs:
https://boisgera.github.io/pandoc/

Don't know how easy it can be to use it in this case

Reply:
Yeh I was hesitant on that one, it doesn't seem to have been widely adopted (by download count (https://pypistats.org/packages/pandoc) and GitHub stars), although it does have recent commits. It is also on a personal account, if the maintainer ever stops maintaining it, we would have to fork it and start to own it ourselves. If this was for a development dependency, it would probably be ok since we could just stop using it. However, this would be a production dependency and possibly have to be used during the build process for sensitive charms exposing us to supply chain attacks. It would also run in an environment where it has access to private API keys for interacting with discourse, opening up an attack vector to someone impersonating Canonical. Based on that, it doesn't seem to be worth it to use the package for a minor reduction in code complexity

Reply:
I follow your opinion, as long as the "markdown parsing" is limited to what we do currently that's still fine, if we start implementing parsing of bold/italic/color/font, ... then we will have to re-evaluate our position :)